### PR TITLE
Allow metrics to be specified as must have data.

### PR DESCRIPTION
**Pull request by fieldju**
While doing some testing I discovered that I wanted the feature described in https://github.com/spinnaker/kayenta/issues/330.

This PR is an attempt at solving that request, hopefully this is something close to what you were thinking about.

### CODE REVIEW

**Comment by fieldju** [comment]
I tested this with my integration test.

Config:
```json
{
      "name": "Number of widgets sold",
      "query": {
        "metricName": "widgets.sold.count",
        "aggregationMethod": "sum",
        "serviceType": "signalfx",
        "type": "signalfx"
      },
      "analysisConfigurations": {
        "canary": {
          "direction": "decrease",
          "nanStrategy": "remove",
          "mustHaveData": true
        }
      },
      "groups": [
        "Super Important Group"
      ],
      "scopeName": "default"
    }
```

Results:
```json
            {
              "name": "Number of widgets sold",
              "tags": {},
              "id": "dcadb8bc-813b-425f-92f9-8b7ab9effca3",
              "classification": "NodataFailMetric",
              "classificationReason": "Missing data for Number of widgets sold",
              "groups": [
                "Super Important Group"
              ],
              "experimentMetadata": {
                "stats": {
                  "std": 0.0,
                  "min": 0.0,
                  "max": 0.0,
                  "mean": 0.0,
                  "count": 0
                }
              },
              "controlMetadata": {
                "stats": {
                  "std": 0.0,
                  "min": 0.0,
                  "max": 0.0,
                  "mean": 0.0,
                  "count": 0
                }
              },
              "resultMetadata": {
                "ratio": 1.0
              },
              "critical": false
            },
```

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MannWhitneyClassifier.scala [hunk 1]
```diff
@@ -128,12 +128,17 @@ class MannWhitneyClassifier(tolerance: Double=0.25,
...
                         nanStrategy: NaNStrategy,
-                        isCriticalMetric: Boolean): MetricClassification = {
+                        isCriticalMetric: Boolean,
+                        isDataRequired: Boolean): MetricClassification = {
 
     //Check if there is no-data for the experiment or control
     if (experiment.values.isEmpty || control.values.isEmpty) {
       if (nanStrategy == NaNStrategy.Remove) {
         val reason = s"Missing data for ${experiment.name}"
+        // if the config indicates that the given metric should have data but not critically fail the canary.
```
Comment (duftler) [inline kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MannWhitneyClassifier.scala]: Missing the word 'Check' from the beginning of the comment (to match the comments above and below).

Could also swap the order of `!isCriticalMetric` and `isDataRequired` in the conditional to match the ordering in the comment.

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MeanInequalityClassifier.scala [hunk 2]
```diff
@@ -31,11 +31,16 @@ class MeanInequalityClassifier extends BaseMetricClassifier {
...
                         direction: MetricDirection,
                         nanStrategy: NaNStrategy,
-                        isCriticalMetric: Boolean): MetricClassification = {
+                        isCriticalMetric: Boolean,
+                        isDataRequired: Boolean): MetricClassification = {
 
     //Check if there is no-data for the experiment or control
     if (experiment.values.isEmpty || control.values.isEmpty) {
       if (nanStrategy == NaNStrategy.Remove) {
+        // if the config indicates that the given metric should have data but not critically fail the canary.
```
Comment (duftler) [inline kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MeanInequalityClassifier.scala]: Same comment as I had above.

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/RandomClassifier.scala [hunk 3]
```diff
@@ -38,11 +38,16 @@ class RandomClassifier(labels: List[MetricClassificationLabel] = List(Pass, High
...
                         direction: MetricDirection,
                         nanStrategy: NaNStrategy,
-                        isCriticalMetric: Boolean): MetricClassification = {
+                        isCriticalMetric: Boolean,
+                        isDataRequired: Boolean): MetricClassification = {
 
     //Check if there is no-data for the experiment or control
     if (experiment.values.isEmpty || control.values.isEmpty) {
       if (nanStrategy == NaNStrategy.Remove) {
+        // if the config indicates that the given metric should have data but not critically fail the canary.
```
Comment (duftler) [inline kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/RandomClassifier.scala]: Same comment as I had above.

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MeanInequalityClassifier.scala (hunk 2 above)
Comment (fieldju) [reply kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MeanInequalityClassifier.scala]: - [X] fixed

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MannWhitneyClassifier.scala (hunk 1 above)
Comment (fieldju) [reply kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/MannWhitneyClassifier.scala]: - [X] fixed

File: kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/RandomClassifier.scala (hunk 3 above)
Comment (fieldju) [reply kayenta-judge/src/main/scala/com/netflix/kayenta/judge/classifiers/metric/RandomClassifier.scala]: - [X] fixed

**Pull request merged**